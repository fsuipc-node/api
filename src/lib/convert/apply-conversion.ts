import { VM } from 'vm2';

import { Offset } from '@shared/offset';
import { RawOffsetValue } from '@shared/offset-values';
import { ConvertedOffsetValue } from '@shared/converted-offset-values';

export const applyConversion = (offset: Offset, rawOffsetValue: RawOffsetValue): ConvertedOffsetValue => {
  if (!offset.convert) {
    return rawOffsetValue;
  }

  if (offset.hasMapping) {
    if (offset.isMappingInvalid) {
      return 'INVALID_MAPPING_FUNCTION';
    }

    return offset.mappingFunction(rawOffsetValue);
  } else {
    // tslint:disable-next-line:no-eval
    if (offset.isInvalidConvertExpression) {
      return 'UNSUPPORTED_CONVERSION_EXPRESSION';
    }

    return new VM().run(
      replaceOffsetExpressionValue(offset, rawOffsetValue)
    );
  }
};

export const replaceOffsetExpressionValue = (offset: Offset, rawOffsetValue: RawOffsetValue): string => {
  return offset.convert.replace(
    new RegExp(/{VAL}/g),
    Array.isArray(rawOffsetValue)
      ? `${JSON.stringify(rawOffsetValue)}`
      : typeof rawOffsetValue === 'string'
        ? `'${rawOffsetValue}'`
        : rawOffsetValue.toString()
  );
};
