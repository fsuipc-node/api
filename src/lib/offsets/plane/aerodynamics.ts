import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const APU: OffsetList = {
  incidenceAlpha: new Offset({
    value: 0x2ED0,
    name: 'incidenceAlpha',
    description: 'incidence alpha (body AoA) - radians',
    category: OffsetCategory.AERODYNAMICS,
    type: Type.Double,
    permission: 'r',
  }),
};
