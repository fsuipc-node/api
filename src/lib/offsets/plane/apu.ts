import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const APU: OffsetList = {
  APUGeneratorSwitch: new Offset({
    value: 0xB51,
    name: 'APUGeneratorSwitch',
    category: OffsetCategory.APU,
    description: 'APU generator switch',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  APUGeneratorActive: new Offset({
    value: 0xB52,
    name: 'APUGeneratorActive',
    category: OffsetCategory.APU,
    description: 'APU generator active',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  APUOnFire: new Offset({
    value: 0xB53,
    name: 'APUOnFire',
    category: OffsetCategory.APU,
    description: 'APU generator active',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  APUMaxRPM: new Offset({
    value: 0xB54,
    name: 'APUMaxRPM',
    category: OffsetCategory.APU,
    description: 'APU generator max RPM percent',
    convert: '+({VAL}).toFixed(2)',
    type: Type.Single,
    permission: 'r',
  }),
  APUStarter: new Offset({
    value: 0xB58,
    name: 'APUStarter',
    category: OffsetCategory.APU,
    description: 'APU starter',
    convert: '{VAL} ? 1 : 0',
    type: Type.Single,
    permission: 'rw',
  }),
  APUVoltage: new Offset({
    value: 0xB5C,
    name: 'APUVoltage',
    category: OffsetCategory.APU,
    description: 'APU generator voltage',
    convert: '+({VAL}).toFixed(2)',
    type: Type.Single,
    permission: 'rw',
  }),
};
