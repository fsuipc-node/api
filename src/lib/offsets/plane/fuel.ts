import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const fuel: OffsetList = {
  fuelWeight: new Offset({
    value: 0xAF4,
    name: 'fuelWeight',
    category: OffsetCategory.FUEL,
    description: 'fuel weight - lbs/gal',
    type: Type.Int16,
    convert: '{VAL} / 256',
    permission: 'r',
  }),
  fuelTankSelector: new Offset({
    value: 0xAF8,
    name: 'fuelTankSelector',
    category: OffsetCategory.FUEL,
    description: '0=Off, 1=All, 2=Left, 3=Right, 4=LeftAux, 5=RightAux, 6=Center, 7=Centre2, 8=Centre3, 9=Ext1, 10=Ext2, 11=RtTip, 12=LftTip, 13=CrossFeed, 14=CrossFeed Left to Right, 15=CrossFeed Right to Left, 16=crossfeed both, 17=external, 18=isolate, 19=left main, 20=right main --- Engine 1 only',
    type: Type.Int16,
    permission: 'rw',
  }),
  fuelCenterTankLevel: new Offset({
    value: 0xB74,
    name: 'fuelCenterTankLevel',
    category: OffsetCategory.FUEL,
    description: 'center tank level - percent',
    convert: 'Math.round({VAL} / (128 * 65536) * 100)',
    type: Type.Int32,
    permission: 'rw',
  }),
  fuelCenterTankCapacity: new Offset({
    value: 0xB78,
    name: 'fuelCenterTankCapacity',
    category: OffsetCategory.FUEL,
    description: 'center tank capacity - US gallons',
    type: Type.Int32,
    permission: 'r',
  }),
  fuelLeftMainTankLevel: new Offset({
    value: 0xB7C,
    name: 'fuelLeftMainTankLevel',
    category: OffsetCategory.FUEL,
    description: 'left main tank level - percent',
    convert: 'Math.round({VAL} / (128 * 65536) * 100)',
    type: Type.Int32,
    permission: 'rw',
  }),
  fuelLeftMainTankCapacity: new Offset({
    value: 0xB80,
    name: 'fuelLeftMainTankCapacity',
    category: OffsetCategory.FUEL,
    description: 'left main tank capacity - US gallons',
    type: Type.Int32,
    permission: 'r',
  }),
  fuelLeftAuxTankLevel: new Offset({
    value: 0xB84,
    name: 'fuelLeftAuxTankLevel',
    category: OffsetCategory.FUEL,
    description: 'left aux tank level - percent',
    convert: 'Math.round({VAL} / (128 * 65536) * 100)',
    type: Type.Int32,
    permission: 'rw',
  }),
  fuelLeftAuxTankCapacity: new Offset({
    value: 0xB88,
    name: 'fuelLeftAuxTankCapacity',
    category: OffsetCategory.FUEL,
    description: 'left aux tank capacity - US gallons',
    type: Type.Int32,
    permission: 'r',
  }),
  fuelLeftTipTankLevel: new Offset({
    value: 0xB8C,
    name: 'fuelLeftTipTankLevel',
    category: OffsetCategory.FUEL,
    description: 'left tip tank level - percent',
    convert: 'Math.round({VAL} / (128 * 65536) * 100)',
    type: Type.Int32,
    permission: 'rw',
  }),
  fuelLeftTipTankCapacity: new Offset({
    value: 0xB90,
    name: 'fuelLeftTipTankCapacity',
    category: OffsetCategory.FUEL,
    description: 'left tip tank capacity - US gallons',
    type: Type.Int32,
    permission: 'r',
  }),
  fuelRightMainTankLevel: new Offset({
    value: 0xB94,
    name: 'fuelRightMainTankLevel',
    category: OffsetCategory.FUEL,
    description: 'right main tank level - percent',
    convert: 'Math.round({VAL} / (128 * 65536) * 100)',
    type: Type.Int32,
    permission: 'rw',
  }),
  fuelRightMainTankCapacity: new Offset({
    value: 0xB98,
    name: 'fuelRightMainTankCapacity',
    category: OffsetCategory.FUEL,
    description: 'right main tank capacity - US gallons',
    type: Type.Int32,
    permission: 'r',
  }),
  fuelRightAuxTankLevel: new Offset({
    value: 0xB9C,
    name: 'fuelRightAuxTankLevel',
    category: OffsetCategory.FUEL,
    description: 'right aux tank level - percent',
    convert: 'Math.round({VAL} / (128 * 65536) * 100)',
    type: Type.Int32,
    permission: 'rw',
  }),
  fuelRightAuxTankCapacity: new Offset({
    value: 0xBA0,
    name: 'fuelRightAuxTankCapacity',
    category: OffsetCategory.FUEL,
    description: 'right aux tank capacity - US gallons',
    type: Type.Int32,
    permission: 'r',
  }),
  fuelRightTipTankLevel: new Offset({
    value: 0xBA4,
    name: 'fuelRightTipTankLevel',
    category: OffsetCategory.FUEL,
    description: 'right tip tank level - percent',
    convert: 'Math.round({VAL} / (128 * 65536) * 100)',
    type: Type.Int32,
    permission: 'rw',
  }),
  fuelRightTipTankCapacity: new Offset({
    value: 0xBA8,
    name: 'fuelRightTipTankCapacity',
    category: OffsetCategory.FUEL,
    description: 'right tip tank capacity - US gallons',
    type: Type.Int32,
    permission: 'r',
  }),
};
