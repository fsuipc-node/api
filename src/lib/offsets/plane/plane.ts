import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const plane: OffsetList = {
  vs0: new Offset({
    value: 0x538,
    name: 'vs0',
    category: OffsetCategory.PLANE,
    description: 'Stall speed full flaps - kt',
    convert: 'ftsecToKt',
    mapping: true,
    type: Type.Double,
    permission: 'r',
  }),
  vs1: new Offset({
    value: 0x540,
    name: 'vs1',
    category: OffsetCategory.PLANE,
    description: 'Stall speed clean - kt',
    convert: 'ftsecToKt',
    mapping: true,
    type: Type.Double,
    permission: 'r',
  }),
  vc: new Offset({
    value: 0x548,
    name: 'vc',
    category: OffsetCategory.PLANE,
    description: 'Cruise speed - kt',
    convert: 'ftsecToKt',
    mapping: true,
    type: Type.Double,
    permission: 'r',
  }),
  vmd: new Offset({
    value: 0x550,
    name: 'vmd',
    category: OffsetCategory.PLANE,
    description: 'Minimum drag speed - kt',
    convert: 'ftsecToKt',
    mapping: true,
    type: Type.Double,
    permission: 'r',
  }),
  smokeSystemControl: new Offset({
    value: 0x5D8,
    name: 'smokeSystemControl',
    category: OffsetCategory.PLANE,
    convert: '!!{VAL}',
    description: 'smoke system control on/off',
    type: Type.UInt16,
    permission: 'rw',
  }),
  retractableGear: new Offset({
    value: 0x60C,
    name: 'retractableGear',
    category: OffsetCategory.PLANE,
    description: 'retractable gear (read only)',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  retractableLeftFloatExtension: new Offset({
    value: 0x614,
    name: 'retractableLeftFloatExtension',
    category: OffsetCategory.PLANE,
    description: 'retractable left float extension - percent',
    convert: 'Math.floor({VAL} / 16384 * 100)',
    type: Type.UInt16,
    permission: 'r',
  }),
  retractableRightFloatExtension: new Offset({
    value: 0x616,
    name: 'retractableRightFloatExtension',
    category: OffsetCategory.PLANE,
    description: 'retractable right float extension - percent',
    convert: 'Math.floor({VAL} / 16384 * 100)',
    type: Type.UInt16,
    permission: 'r',
  }),
  hasFlaps: new Offset({
    value: 0x778,
    name: 'hasFlaps',
    category: OffsetCategory.PLANE,
    description: 'has flaps',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasStallHorn: new Offset({
    value: 0x77C,
    name: 'hasStallHorn',
    category: OffsetCategory.PLANE,
    description: 'has stall horn',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasMixtureEngine: new Offset({
    value: 0x780,
    name: 'hasMixtureEngine',
    category: OffsetCategory.PLANE,
    description: 'has mixture engine',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasCarbHeat: new Offset({
    value: 0x784,
    name: 'hasCarbHeat',
    category: OffsetCategory.PLANE,
    description: 'has carburator heater',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasSpoilers: new Offset({
    value: 0x78C,
    name: 'hasSpoilers',
    category: OffsetCategory.PLANE,
    description: 'has spoilers',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  isTailDragger: new Offset({
    value: 0x790,
    name: 'isTailDragger',
    category: OffsetCategory.PLANE,
    description: 'has spoilers',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasStrobes: new Offset({
    value: 0x794,
    name: 'hasStrobes',
    category: OffsetCategory.PLANE,
    description: 'has strobes',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasToeBrakes: new Offset({
    value: 0x79C,
    name: 'hasToeBrakes',
    category: OffsetCategory.PLANE,
    description: 'has toe brakes',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  maxMach: new Offset({
    value: 0xB0C,
    name: 'maxMach',
    category: OffsetCategory.PLANE,
    description: 'mach maximum operating - not FS2004',
    convert: '+({VAL} / 20438).toFixed(2)',
    type: Type.Int32,
    permission: 'r',
  }),
  fuelSelectorsCount: new Offset({
    value: 0x123E,
    name: 'fuelSelectorsCount',
    category: OffsetCategory.PLANE,
    description: 'number of fuel selectors for the aircraft',
    type: Type.Byte,
    permission: 'r',
  }),
  aircraftEmptyWeight: new Offset({
    value: 0x1330,
    name: 'aircraftEmptyWeight',
    category: OffsetCategory.PLANE,
    description: 'aicraft empty weight',
    convert: '+({VAL} / 256).toFixed(2)',
    type: Type.Int32,
    permission: 'r',
  }),
  aircraftMaxGrossWeight: new Offset({
    value: 0x1334,
    name: 'aircraftMaxGrossWeight',
    category: OffsetCategory.PLANE,
    description: 'aicraft max gross weight',
    convert: '+({VAL} / 256).toFixed(2)',
    type: Type.Int32,
    permission: 'r',
  }),
};
