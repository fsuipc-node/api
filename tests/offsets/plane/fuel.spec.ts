import { fuel as offsets } from '@offsets/plane/fuel';
import { replaceOffsetExpressionValue } from '@convert/apply-conversion';

describe('offset - plane/fuel', () => {
  const offsetsTestCases = [
    { name: 'fuelWeight', value: 512, expectedResult: 2 },
    { name: 'fuelCenterTankLevel', value: 4194304, expectedResult: 50 },
    { name: 'fuelLeftMainTankLevel', value: 4194304, expectedResult: 50 },
    { name: 'fuelLeftAuxTankLevel', value: 4194304, expectedResult: 50 },
    { name: 'fuelLeftTipTankLevel', value: 4194304, expectedResult: 50 },
    { name: 'fuelRightMainTankLevel', value: 4194304, expectedResult: 50 },
    { name: 'fuelRightAuxTankLevel', value: 4194304, expectedResult: 50 },
    { name: 'fuelRightTipTankLevel', value: 4194304, expectedResult: 50 },
  ];

  describe('offsets list', () => {
    it('should have required properties', () => {
      expect(offsets).toMatchSnapshot();
    });
  });

  offsetsTestCases.forEach(testedOffset => {
    describe(testedOffset.name, () => {
      it('should convert data properly', () => {
        const convertExpression = replaceOffsetExpressionValue(offsets[testedOffset.name], testedOffset.value);

        // tslint:disable-next-line:no-eval
        expect(eval(convertExpression)).toEqual(testedOffset.expectedResult);
      });
    });
  });
});
