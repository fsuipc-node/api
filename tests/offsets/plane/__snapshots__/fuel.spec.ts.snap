// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`offset - plane/fuel offsets list should have required properties 1`] = `
Object {
  "fuelCenter2TankCapacity": Offset {
    "category": "fuel",
    "convert": undefined,
    "description": "center 2 tank capacity - US gallons",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelCenter2TankCapacity",
    "permission": "r",
    "type": 3,
    "value": 4680,
  },
  "fuelCenter2TankLevel": Offset {
    "category": "fuel",
    "convert": "Math.round({VAL} / (128 * 65536) * 100)",
    "description": "center 2 tank level - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelCenter2TankLevel",
    "permission": "rw",
    "type": 3,
    "value": 4676,
  },
  "fuelCenter3TankCapacity": Offset {
    "category": "fuel",
    "convert": undefined,
    "description": "center 3 tank capacity - US gallons",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelCenter3TankCapacity",
    "permission": "r",
    "type": 3,
    "value": 4688,
  },
  "fuelCenter3TankLevel": Offset {
    "category": "fuel",
    "convert": "Math.round({VAL} / (128 * 65536) * 100)",
    "description": "center 3 tank level - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelCenter3TankLevel",
    "permission": "rw",
    "type": 3,
    "value": 4684,
  },
  "fuelCenterTankCapacity": Offset {
    "category": "fuel",
    "convert": undefined,
    "description": "center tank capacity - US gallons",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelCenterTankCapacity",
    "permission": "r",
    "type": 3,
    "value": 2936,
  },
  "fuelCenterTankLevel": Offset {
    "category": "fuel",
    "convert": "Math.round({VAL} / (128 * 65536) * 100)",
    "description": "center tank level - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelCenterTankLevel",
    "permission": "rw",
    "type": 3,
    "value": 2932,
  },
  "fuelExternal1TankCapacity": Offset {
    "category": "fuel",
    "convert": undefined,
    "description": "external 1 tank capacity - US gallons",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelExternal1TankCapacity",
    "permission": "r",
    "type": 3,
    "value": 4696,
  },
  "fuelExternal1TankLevel": Offset {
    "category": "fuel",
    "convert": "Math.round({VAL} / (128 * 65536) * 100)",
    "description": "external 1 tank level - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelExternal1TankLevel",
    "permission": "rw",
    "type": 3,
    "value": 4692,
  },
  "fuelExternal2TankCapacity": Offset {
    "category": "fuel",
    "convert": undefined,
    "description": "external 2 tank capacity - US gallons",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelExternal2TankCapacity",
    "permission": "r",
    "type": 3,
    "value": 4704,
  },
  "fuelExternal2TankLevel": Offset {
    "category": "fuel",
    "convert": "Math.round({VAL} / (128 * 65536) * 100)",
    "description": "external 2 tank level - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelExternal2TankLevel",
    "permission": "rw",
    "type": 3,
    "value": 4700,
  },
  "fuelLeftAuxTankCapacity": Offset {
    "category": "fuel",
    "convert": undefined,
    "description": "left aux tank capacity - US gallons",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelLeftAuxTankCapacity",
    "permission": "r",
    "type": 3,
    "value": 2952,
  },
  "fuelLeftAuxTankLevel": Offset {
    "category": "fuel",
    "convert": "Math.round({VAL} / (128 * 65536) * 100)",
    "description": "left aux tank level - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelLeftAuxTankLevel",
    "permission": "rw",
    "type": 3,
    "value": 2948,
  },
  "fuelLeftMainTankCapacity": Offset {
    "category": "fuel",
    "convert": undefined,
    "description": "left main tank capacity - US gallons",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelLeftMainTankCapacity",
    "permission": "r",
    "type": 3,
    "value": 2944,
  },
  "fuelLeftMainTankLevel": Offset {
    "category": "fuel",
    "convert": "Math.round({VAL} / (128 * 65536) * 100)",
    "description": "left main tank level - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelLeftMainTankLevel",
    "permission": "rw",
    "type": 3,
    "value": 2940,
  },
  "fuelLeftTipTankCapacity": Offset {
    "category": "fuel",
    "convert": undefined,
    "description": "left tip tank capacity - US gallons",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelLeftTipTankCapacity",
    "permission": "r",
    "type": 3,
    "value": 2960,
  },
  "fuelLeftTipTankLevel": Offset {
    "category": "fuel",
    "convert": "Math.round({VAL} / (128 * 65536) * 100)",
    "description": "left tip tank level - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelLeftTipTankLevel",
    "permission": "rw",
    "type": 3,
    "value": 2956,
  },
  "fuelRightAuxTankCapacity": Offset {
    "category": "fuel",
    "convert": undefined,
    "description": "right aux tank capacity - US gallons",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelRightAuxTankCapacity",
    "permission": "r",
    "type": 3,
    "value": 2976,
  },
  "fuelRightAuxTankLevel": Offset {
    "category": "fuel",
    "convert": "Math.round({VAL} / (128 * 65536) * 100)",
    "description": "right aux tank level - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelRightAuxTankLevel",
    "permission": "rw",
    "type": 3,
    "value": 2972,
  },
  "fuelRightMainTankCapacity": Offset {
    "category": "fuel",
    "convert": undefined,
    "description": "right main tank capacity - US gallons",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelRightMainTankCapacity",
    "permission": "r",
    "type": 3,
    "value": 2968,
  },
  "fuelRightMainTankLevel": Offset {
    "category": "fuel",
    "convert": "Math.round({VAL} / (128 * 65536) * 100)",
    "description": "right main tank level - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelRightMainTankLevel",
    "permission": "rw",
    "type": 3,
    "value": 2964,
  },
  "fuelRightTipTankCapacity": Offset {
    "category": "fuel",
    "convert": undefined,
    "description": "right tip tank capacity - US gallons",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelRightTipTankCapacity",
    "permission": "r",
    "type": 3,
    "value": 2984,
  },
  "fuelRightTipTankLevel": Offset {
    "category": "fuel",
    "convert": "Math.round({VAL} / (128 * 65536) * 100)",
    "description": "right tip tank level - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelRightTipTankLevel",
    "permission": "rw",
    "type": 3,
    "value": 2980,
  },
  "fuelSelectedQuantity": Offset {
    "category": "fuel",
    "convert": undefined,
    "description": "Selected fuel quantity - US gallons",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelSelectedQuantity",
    "permission": "r",
    "type": 3,
    "value": 4712,
  },
  "fuelTankSelector": Offset {
    "category": "fuel",
    "convert": "fuelTank",
    "description": "Engine 1 only",
    "length": undefined,
    "mapping": true,
    "name": "fuelTankSelector",
    "permission": "rw",
    "type": 2,
    "value": 2808,
  },
  "fuelTotalQuantity": Offset {
    "category": "fuel",
    "convert": undefined,
    "description": "total fuel quantity - US gallons",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelTotalQuantity",
    "permission": "r",
    "type": 3,
    "value": 4708,
  },
  "fuelTotalQuantityWeight": Offset {
    "category": "fuel",
    "convert": undefined,
    "description": "total fuel quantity weight - lb",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelTotalQuantityWeight",
    "permission": "r",
    "type": 3,
    "value": 4716,
  },
  "fuelWeight": Offset {
    "category": "fuel",
    "convert": "{VAL} / 256",
    "description": "fuel weight - lbs/gal",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelWeight",
    "permission": "r",
    "type": 2,
    "value": 2804,
  },
  "totalFuelCapacity": Offset {
    "category": "fuel",
    "convert": undefined,
    "description": "total fuel capacity - US gallons",
    "length": undefined,
    "mapping": undefined,
    "name": "totalFuelCapacity",
    "permission": "r",
    "type": 6,
    "value": 4672,
  },
}
`;
